ENG-5771: Add firebase app distribution support

seanirby opened the pull request:
## Describe your changes

This PR updates the Fastlane plugin to optionally support a configuration for firebase app distribution.  If the config is provided the plugin will add the necessary lanes.  Besides including this configuration, the only other step for projects to integrate with FAD is to update their Github actions/workflows and to create a secret in their repo, GOOGLE_SERVICE_CREDENTIALS, which contains the firebase credentials for the google service account.

Overview of changes:
 - Update Fastfile templates to provide conditional firebase lanes.
 - Update Pluginfile templates to conditionally provide the necessary appcenter and/or firebase gems.
 - Fastlane Pluginfiles templates are now rendered to account for conditional gem installations
 - Updated tests

~~NOTE: I still need to verify Firebase Android release builds. I'm getting a failure on those builds because the app isn't linked to a Google Play account.~~

Firebase release uploads have been verified to be working with this plugin logic using the theory app.

![Screenshot 2024-11-04 at 10 51 05 AM](https://github.com/user-attachments/assets/ec851728-9ce9-4406-b5a9-97e8d28c3b1f)

https://console.firebase.google.com/project/theory-57e35/appdistribution/app/android:com.fastretailing.theory/releases

## Issue ticket number and link

https://brandingbrand.atlassian.net/browse/ENG-5771

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Test Plan

I created an example shopify template using the flagship accelerator. I used the config provided below and updated my repo to include the google service credential secret for the shopify project. I've given Nick Burkhart view permissions to the project so he can check it out. Currently the firebase builds are working for both platforms except the release Android builds aren't uploading because there isn't a Google Play account linked to the Firebase Shopify app.

```
  codePluginFastlane: {
    plugin: {
      ios: {
        appCenter: {
          organization: 'Branding-Brand',
          appName: 'Shopify-iOS-Internal',
          destinationType: 'group',
          destinations: ['IAT'],
        },
        firebase: {
          groups: ['IAT'],
          appId: '1:300374277703:ios:cdf8a81375502895bcbb62',
        },
      },
      android: {
        appCenter: {
          organization: 'Branding-Brand',
          appName:
            'Shopify-Android-Internal',
          destinationType: 'group',
          destinations: ['IAT'],
        },
        firebase: {
          groups: ['IAT'],
          appId: '1:300374277703:android:764c235c8f6d7dc6bcbb62',
        },
      },
    },
  },
```

## Checklist before requesting a review

- [x] A self-review of my code has been completed
- [x] Tests have been added / updated if required
- [x] Documentation has been updated to reflect these changes

NickBurkhartBB commented:
Note: as currently setup in PR this is a breaking change as it changes the lanes you use to upload.

NickBurkhartBB commented on packages/plugin-fastlane/template/android/fastlane/Fastfile:
```diff
@@ -10,53 +10,90 @@ lane :bundle do
+<% } -%>
+end
 
+lane :distribute_store do
```
nitpick: rather then distribute_[store|internal], I think it might make sense to have the name reflect the output type closer to android naming, so distribute_[package|bundle]

seanirby replied:
Sure

seanirby commented on packages/plugin-fastlane/src/index.ts:
```diff
@@ -78,6 +78,18 @@ eval_gemfile(plugins_path) if File.exist?(plugins_path)`,
 
+    // Render Pluginfile template for iOS
+    await withUTF8(
+      path.project.resolve('ios', 'fastlane', 'Pluginfile'),
```
Yeah this was a leftover from when I initially had the gems conditional. I decided to revert back to including them both unconditionally in case a project wanted to increment their firebase build number off the appcenter version.

seanirby commented:
I've addressed PR comments and also updated the lanes so this update should be nonbreaking.  I did a quick test on theory repo with the existing workflows: https://github.com/brandingbrand/theory/actions/runs/11692375987

NickBurkhartBB reviewed (changes requested):
@seanirby It looks good, just a nitpick about the outputted Fastfile, the whitespace is a little messed up due to extra indents in the template due to indenting based off EJS if statements.

seanirby commented:
> @seanirby It looks good, just a nitpick about the outputted Fastfile, the whitespace is a little messed up due to extra indents in the template due to indenting based off EJS if statements.

I'll fix this up.

NickBurkhartBB commented on packages/plugin-fastlane/template/android/fastlane/Fastfile:
```diff
@@ -31,6 +31,37 @@ lane :increment_build do
 <% } -%>
 
+lane :increment_build_appcenter do
+  <% if(codePluginFastlane.plugin && codePluginFastlane.plugin.android && codePluginFastlane.plugin.android.appCenter) { -%>
```
@seanirby sorry to be a bother but the whitespace before the if/else statements get included in the output as well. since we don't want any whitespace on EJS lines you can move them all the way to the left. Can you double check output on iOS and Android as I am still seeing these issues on both.

seanirby replied:
Sorry about that, I'll fix this.
